core/embed/rust/src/trezorhal/bitblt.rs

```diff
@@ -0,0 +1,216 @@
+        Self {
+            dst_row: unsafe { dst.row_ptr(self.dst_y) },
+            dst_stride: dst.stride() as u16,
+            ..self
+        }
+    }
+
+    fn with_rect(self, r: Rect) -> Self {
+        Self {
+            width: r.width() as u16,
+            height: r.height() as u16,
+            dst_x: r.x0 as u16,
+            dst_y: r.y0 as u16,
+            ..self
+        }
+    }
+
+    fn with_src(self, bitmap: &Bitmap, x: i16, y: i16) -> Self {
+        let bitmap_stride = match bitmap.format() {
+            BitmapFormat::MONO1P => bitmap.width() as u16, // packed bits
+            _ => bitmap.stride() as u16,
+        };
+
+        Self {
+            src_row: unsafe { bitmap.row_ptr(y as u16) },
+            src_stride: bitmap_stride,
+            src_x: x as u16,
+            src_y: y as u16,
+            ..self
+        }
+    }
+
+    fn with_fg(self, fg_color: Color) -> Self {
+        Self {
+            src_fg: fg_color.into(),
+            ..self
+        }
+    }
+
+    fn with_bg(self, bg_color: Color) -> Self {
+        Self {
+            src_bg: bg_color.into(),
+            ..self
+        }
+    }
+
+    fn with_alpha(self, alpha: u8) -> Self {
+        Self {
+            src_alpha: alpha,
+            ..self
+        }
+    }
+
+    pub fn wait_for_transfer() {
+        #[cfg(feature = "dma2d")]
+        unsafe {
+            ffi::dma2d_wait()
+        }
+    }
+
+    pub unsafe fn rgb565_fill(&self) {
+        unsafe { ffi::gl_rgb565_fill(self) };
+    }
+
+    pub unsafe fn rgb565_copy_mono4(&self) {
+        unsafe { ffi::gl_rgb565_copy_mono4(self) };
+    }
+
+    pub unsafe fn rgb565_copy_rgb565(&self) {
+        unsafe { ffi::gl_rgb565_copy_rgb565(self) };
+    }
+
+    pub unsafe fn rgb565_blend_mono4(&self) {
+        unsafe { ffi::gl_rgb565_blend_mono4(self) };
+    }
+
+    pub unsafe fn rgba8888_fill(&self) {
+        unsafe { ffi::gl_rgba8888_fill(self) };
+    }
+
+    pub unsafe fn rgba8888_copy_mono4(&self) {
+        unsafe { ffi::gl_rgba8888_copy_mono4(self) };
+    }
+
+    pub unsafe fn rgba8888_copy_rgb565(&self) {
+        unsafe { ffi::gl_rgba8888_copy_rgb565(self) };
+    }
+
+    pub unsafe fn rgba8888_copy_rgba8888(&self) {
+        unsafe { ffi::gl_rgba8888_copy_rgba8888(self) };
+    }
+
+    pub unsafe fn rgba8888_blend_mono4(&self) {
+        unsafe { ffi::gl_rgba8888_blend_mono4(self) };
+    }
+
+    pub unsafe fn mono8_fill(&self) {
+        unsafe { ffi::gl_mono8_fill(self) };
+    }
+
+    pub unsafe fn mono8_copy_mono1p(&self) {
+        unsafe { ffi::gl_mono8_copy_mono1p(self) };
+    }
+
+    pub unsafe fn mono8_copy_mono4(&self) {
+        unsafe { ffi::gl_mono8_copy_mono4(self) };
+    }
+
+    pub unsafe fn mono8_blend_mono1p(&self) {
+        unsafe { ffi::gl_mono8_blend_mono1p(self) };
+    }
+
+    pub unsafe fn mono8_blend_mono4(&self) {
+        unsafe { ffi::gl_mono8_blend_mono4(self) };
+    }
+
+    #[cfg(feature = "new_rendering")]
```

Comment:
this and the `display_copy_rgb565` function should only be compiled if not using `xframebuffer`

Reply:
Yes, it could be. However, it still has to be compiled only if new_rendering is on. Fortunately, this complicated condition is temporary.

see c1c7f3e659532f91f3466e5f523260f5160b617d